devtools/working-with-dsproxy/working-with-dsproxy.md

```diff
@@ -0,0 +1,437 @@
+
+`msg.sender` when the script is being executed will continue to be the user address instead of the address of the DSProxy contract.
+
+`execute(bytes code, bytes data)` is an additional function that can be used when a user wants to deploy a contract containing scripts and then call one of the scripts in a single transaction. A `cache` registers the address of contract deployed to save gas by skipping deployment when other users call `execute` with the same bytecode later.
+
+#### Event Logs
+
+A DSProxy contract generates a event called `LogNote` with these values indexed when `execute()` is called,
+
+* Function signature, `0x1cff79cd`
+* Owner of the DSProxy contract, `msg.sender`
+* Contract address which contains the script, `address`
+* Calldata which contains function signature of script being executed and its input data, `data`
+
+#### Factory Contract
+
+The function  `build` in the DSProxyFactory contract is used to deploy a personal DSProxy contract. For production usecases on mainnet you can use a common factory contract that is already being used by existing projects to avoid deploying redundant DSProxy contracts for users who already have one. Please check the [Production Usage](/production-usage) section in this guide for more details.
+
+### Create a script
+
+We've seen an example earlier of how a script can help CDP owners pay back their Dai debt by purchasing the required amount of MKR from an exchange within the same transaction. Uniswap exchange contracts are a good source of liquidity especially for buying small amounts of MKR. In this section, we will create a script that will allow users to buy MKR with DAI from Uniswap and wipe debt from a CDP.
+
+#### Environment Setup
+
+We'll use `dapp` and `seth` while working through this section but you can also your own tool of choice like the Remix IDE to execute these steps. Instructions to install both the tools can be found [here](https://dapp.tools/).
+
+You have to create a `~/.sethrc` file and configure it with these values to work with the Kovan testnet,
+
+* `SETH_CHAIN=kovan`
+* `ETH_FROM=0xYourKovanAddressFromKeyStoreOrLedger`
+* `ETH_GAS=4000000`
+* `ETH_GAS_PRICE=2500000000`
+
+Seth uses an Infura RPC URL by default but you can also configure `ETH_RPC_URL` and point it to your preferred end-point.
+
+#### Create a new dapp project
+
+Create a new folder and open it
+
+```bash
+mkdir wipe-proxy && cd wipe-proxy
+```
+
+Initialize a dapp project within it
+
+```bash
+dapp init
+```
+
+Add the DSMath library to the project to use safe math operations within the script
+
+```text
+dapp install ds-math
+```
+
+#### Setup WipeProxy.sol
+
+Import the DSMath contract and add it to the WipeProxy contract
+
+```text
+import "ds-math/math.sol";
+
```

Comment:
Probably because I'm doing this late at night.. I initially thought to include all the interfaces in the `WipeProxy` Contract. This same mistake by others may be avoided if the interfaces are described right after `import "ds-math/math.sol";`

Reply:
I split this into two separate steps and moved them around to reflect the order in which code is added to the contract.